src/tlsdata.c

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright (C) the libgit2 contributors. All rights reserved.
+ *
+ * This file is part of libgit2, distributed under the GNU GPL v2 with
+ * a Linking Exception. For full terms see the included COPYING file.
+ */
+
+#include "tlsdata.h"
+#include "global.h"
+
+static void tlsdata_dispose(git_tlsdata *tlsdata);
+
+/**
+ * Handle the thread-local state
+ *
+ * `git_tlsdata_global_init` will be called as part
+ * of `git_libgit2_init` (which itself must be called
+ * before calling any other function in the library).
+ *
+ * This function allocates a TLS index (using pthreads
+ * or fiber-local storage in Win32) to store the per-
+ * thread state.
+ *
+ * Any internal method that requires thread-local state
+ * will then call `git_tlsdata_get()` which returns a
+ * pointer to the thread-local state structure; this
+ * structure is lazily allocated on each thread.
+ *
+ * This mechanism will register a shutdown handler
+ * (`git_tlsdata_global_shutdown`) which will free the
+ * TLS index.  This shutdown handler will be called by
+ * `git_libgit2_shutdown`.
+ *
+ * If libgit2 is built without threading support, the
+ * `git_tlsdata_get()` call returns a pointer to a single,
+ * statically allocated global state. The `git_thread_`
+ * functions are not available in that case.
+ */
+
+#if defined(GIT_THREADS) && defined(GIT_WIN32)
+
+static DWORD fls_index;
+
+static void git_tlsdata_global_shutdown(void)
+{
+	FlsFree(fls_index);
+}
+
+static void WINAPI fls_free(void *tlsdata)
+{
+	tlsdata_dispose(tlsdata);
+	git__free(tlsdata);
+}
+
+int git_tlsdata_global_init(void)
+{
+	if ((fls_index = FlsAlloc(fls_free)) == FLS_OUT_OF_INDEXES)
+		return -1;
+
+	git__on_shutdown(git_tlsdata_global_shutdown);
+
+	return 0;
+}
+
+git_tlsdata *git_tlsdata_get(void)
+{
+	git_tlsdata *tlsdata;
+
+	if ((tlsdata = FlsGetValue(fls_index)) != NULL)
+		return tlsdata;
+
+	if ((tlsdata = git__calloc(1, sizeof(git_tlsdata))) == NULL)
+		return NULL;
+
+	git_buf_init(&tlsdata->error_buf, 0);
```

Comment:
`git_buf_init` may return an error, even though it won't if used with `initial_size == 0`. I'd still argue we should check for errors.

Reply:
Fixed.